orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc

```diff
@@ -0,0 +1,205 @@
+      }
+      recomputed_inputs.push_back(node->MutableOutputDefs()[1]);
+      recomputed_inputs.push_back(node->MutableInputDefs()[1]);
+      recomputed_inputs.push_back(node->MutableInputDefs()[2]);
+
+      const auto& output = node->OutputDefs()[0];
+      auto& recomputed_output = graph.GetOrCreateNodeArg(graph_utils::RecomputeName(output->Name()),
+                                                         output->TypeAsProto());
+
+      Node& recompute_node = graph.AddNode(node->Name() + "_recompute",
+                                           "DropoutGrad",
+                                           "Recompute of " + node->Name(),
+                                           recomputed_inputs,
+                                           {&recomputed_output},
+                                           {},
+                                           kMSDomain);
+      recompute_node.SetPriority(priority);
+      continue;
+    }
+
+    // prepare inputs for recompute node
+    std::vector<NodeArg*> recomputed_inputs;
+    for (NodeArg* input : node->MutableInputDefs()) {
+      const Node* p_node = graph.GetProducerNode(input->Name());
+
+      // do not duplicate initializers in recompute subgraph
+      if (initializers.find(input->Name()) != initializers.end() ||
+          std::find(nodes.begin(), nodes.end(), p_node) == nodes.end()) {
+        recomputed_inputs.push_back(input);
+      } else {
+        auto& recomputed_input = graph.GetOrCreateNodeArg(graph_utils::RecomputeName(input->Name()),
+                                                          input->TypeAsProto());
+        recomputed_inputs.push_back(&recomputed_input);
+      }
+    }
+
+    // prepare ouputs for recompute node
+    std::vector<NodeArg*> recomputed_outputs;
+    for (NodeArg* output : node->MutableOutputDefs()) {
+      auto& recomputed_output = graph.GetOrCreateNodeArg(graph_utils::RecomputeName(output->Name()),
+                                                         output->TypeAsProto());
+      recomputed_outputs.push_back(&recomputed_output);
+    }
+
+    Node& recompute_node = graph.AddNode(node->Name() + "_recompute",
+                                         node->OpType(),
+                                         "Recompute of " + node->Name(),
+                                         recomputed_inputs,
+                                         recomputed_outputs,
+                                         &node->GetAttributes(),
+                                         node->Domain());
+    recompute_node.SetPriority(priority);
+  }
+  return;
+}
+
+Status TransformerLayerRecompute::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/, const logging::Logger& logger) const {
+  std::vector<std::pair<const NodeArg*, const NodeArg*>> start_end_edges;
+  ORT_RETURN_IF_ERROR(IdentifyTransformerLayerEdges(graph, start_end_edges, logger));
+
+  // insert recompute nodes expect for the last transformer layer
```

Comment:
>insert recompute nodes expect for the last transformer layer [](start = 5, length = 60)

can we have an option to do m-layer of recompute, instead of a fixed n-1?

Reply:
Will do in a future PR when integrate with frontend.